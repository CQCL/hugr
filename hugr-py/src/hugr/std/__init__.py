"""Types and operations from the standard extension set."""

import pkgutil

from hugr._serialization.extension import Extension as PdExtension
from hugr.ext import Extension


def _load_extension(name: str) -> Extension:
    replacement = name.replace(".", "/")
    json_str = pkgutil.get_data(__name__, f"_json_defs/{replacement}.json")
    assert json_str is not None
    return PdExtension.model_validate_json(json_str).deserialize()


PRELUDE = _load_extension("prelude")
