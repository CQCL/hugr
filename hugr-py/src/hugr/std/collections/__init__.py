"""Standard extensions for collection types and operations."""
