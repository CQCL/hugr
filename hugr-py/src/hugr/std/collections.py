"""Collection types and operations."""

from __future__ import annotations

from dataclasses import dataclass

import hugr.tys as tys
from hugr import val
from hugr.std import _load_extension
from hugr.utils import comma_sep_str

EXTENSION = _load_extension("collections")


def list_type(ty: tys.Type) -> tys.ExtType:
    """Returns a list type with a fixed element type."""
    arg = tys.TypeTypeArg(ty)
    return EXTENSION.types["List"].instantiate([arg])


@dataclass
class ListVal(val.ExtensionValue):
    """Constant value for a list of elements."""

    v: list[val.Value]
    ty: tys.Type

    def __init__(self, v: list[val.Value], elem_ty: tys.Type) -> None:
        self.v = v
        self.ty = list_type(elem_ty)

    def to_value(self) -> val.Extension:
        name = "ListValue"
        # The value list must be serialized at this point, otherwise the
        # `Extension` value would not be serializable.
        vs = [v._to_serial_root() for v in self.v]
        return val.Extension(name, typ=self.ty, val=vs, extensions=[EXTENSION.name])

    def __str__(self) -> str:
        return f"[{comma_sep_str(self.v)}]"
