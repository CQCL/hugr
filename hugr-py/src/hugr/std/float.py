"""Floating point types and operations."""

from __future__ import annotations

from dataclasses import dataclass

from hugr import val
from hugr.std import _load_extension

FLOAT_TYPES_EXTENSION = _load_extension("arithmetic.float.types")

FLOAT_T = FLOAT_TYPES_EXTENSION.types["float64"].instantiate([])


FLOAT_OPS_EXTENSION = _load_extension("arithmetic.float")


@dataclass
class FloatVal(val.ExtensionValue):
    """Custom value for a floating point number."""

    v: float

    def to_value(self) -> val.Extension:
        name = "ConstF64"
        payload = {"value": self.v}
        return val.Extension(
            name, typ=FLOAT_T, val=payload, extensions=[FLOAT_TYPES_EXTENSION.name]
        )

    def __str__(self) -> str:
        return f"{self.v}"
