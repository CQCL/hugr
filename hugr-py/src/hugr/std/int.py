"""HUGR integer types and operations."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, ClassVar

from typing_extensions import Self

from hugr import ext, tys, val
from hugr.ops import AsExtOp, DataflowOp, ExtOp, RegisteredOp
from hugr.std import _load_extension

if TYPE_CHECKING:
    from hugr.ops import Command, ComWire

CONVERSIONS_EXTENSION = _load_extension("arithmetic.conversions")

INT_TYPES_EXTENSION = _load_extension("arithmetic.int.types")
_INT_PARAM = tys.BoundedNatParam(7)

INT_T_DEF = INT_TYPES_EXTENSION.types["int"]


def int_t(width: int) -> tys.ExtType:
    """Create an integer type with a fixed log bit width.


    Args:
        width: The log bit width of the integer.

    Returns:
        The integer type.

    Examples:
        >>> int_t(5).type_def.name # 32 bit integer
        'int'
    """
    return INT_T_DEF.instantiate(
        [tys.BoundedNatArg(n=width)],
    )


def _int_tv(index: int) -> tys.ExtType:
    return INT_T_DEF.instantiate(
        [tys.VariableArg(idx=index, param=_INT_PARAM)],
    )


#: HUGR 32-bit integer type.
INT_T = int_t(5)


@dataclass
class IntVal(val.ExtensionValue):
    """Custom value for an integer."""

    v: int
    width: int = field(default=5)

    def to_value(self) -> val.Extension:
        name = "ConstInt"
        payload = {"log_width": self.width, "value": self.v}
        return val.Extension(
            name,
            typ=int_t(self.width),
            val=payload,
            extensions=[INT_TYPES_EXTENSION.name],
        )

    def __str__(self) -> str:
        return f"{self.v}"


INT_OPS_EXTENSION = _load_extension("arithmetic.int")


@dataclass(frozen=True)
class _DivModDef(RegisteredOp):
    """DivMod operation, has two inputs and two outputs."""

    width: int = 5
    const_op_def: ClassVar[ext.OpDef] = INT_OPS_EXTENSION.operations["idivmod_u"]

    def type_args(self) -> list[tys.TypeArg]:
        return [tys.BoundedNatArg(n=self.width)]

    def cached_signature(self) -> tys.FunctionType | None:
        row: list[tys.Type] = [int_t(self.width)] * 2
        return tys.FunctionType.endo(row)

    @classmethod
    def from_ext(cls, custom: ExtOp) -> Self | None:
        if custom.op_def() != cls.op_def():
            return None
        match custom.args:
            case [tys.BoundedNatArg(n=a1)]:
                return cls(width=a1)
            case _:
                msg = f"Invalid args: {custom.args}"
                raise AsExtOp.InvalidExtOp(msg)

    def __call__(self, a: ComWire, b: ComWire) -> Command:
        return DataflowOp.__call__(self, a, b)


#: DivMod operation.
DivMod = _DivModDef()
