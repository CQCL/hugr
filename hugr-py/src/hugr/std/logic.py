"""HUGR logic operations."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, ClassVar

from hugr.ops import Command, DataflowOp, RegisteredOp
from hugr.std import _load_extension

if TYPE_CHECKING:
    from hugr import ext
    from hugr.ops import ComWire


EXTENSION = _load_extension("logic")


@dataclass(frozen=True)
class _NotOp(RegisteredOp):
    """Logical NOT operation."""

    const_op_def: ClassVar[ext.OpDef] = EXTENSION.operations["Not"]

    def __call__(self, a: ComWire) -> Command:
        return DataflowOp.__call__(self, a)


#: Not operation
Not = _NotOp()
