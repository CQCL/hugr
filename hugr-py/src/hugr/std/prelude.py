"""HUGR prelude values."""

from __future__ import annotations

from dataclasses import dataclass

from hugr import val
from hugr.std import _load_extension

PRELUDE_EXTENSION = _load_extension("prelude")

STRING_T_DEF = PRELUDE_EXTENSION.types["string"]

STRING_T = STRING_T_DEF.instantiate([])


@dataclass
class StringVal(val.ExtensionValue):
    """Custom value for a string."""

    v: str

    def to_value(self) -> val.Extension:
        name = "ConstString"
        payload = {"value": self.v}
        return val.Extension(
            name,
            typ=STRING_T,
            val=payload,
            extensions=[PRELUDE_EXTENSION.name],
        )

    def __str__(self) -> str:
        return f"{self.v}"
