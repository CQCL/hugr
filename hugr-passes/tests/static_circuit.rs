use std::{
    fs, path::Path, process::Command,
};

use hugr_core::{
    extension::ExtensionSet, hugr::IdentList, ops::NamedOp, package::Package,
    std_extensions::STD_REG, Hugr, HugrView,
};
use hugr_passes::{
    inline::InlinePass,
    static_circuit::{Gate, StaticCircuitContext, StaticCircuitPass, StaticCircuitResult},
};
use insta::{self, glob};
use itertools::Itertools;

struct T {
    file: Box<Path>,
}

impl T {
    fn new(path: impl AsRef<Path>) -> Self {
        Self {
            file: path.as_ref().into(),
        }
    }

    fn run(&self) -> Result<Package, Box<dyn std::error::Error>> {
        let guppy_output = {
            let mut cmd = Command::new("python3");
            cmd.arg(self.file.as_ref());
            eprintln!("{cmd:?}");
            cmd.output()?
        };

        let status = guppy_output.status;
        if !status.success() {
            Err(format!(
                "Guppy failed for {}: {status}\n{}",
                self.file.as_ref().to_string_lossy(),
                String::from_utf8(guppy_output.stderr).unwrap()
            ))?;
        }

        Package::from_json_reader(guppy_output.stdout.as_slice()).map_err(Into::into)
    }
}

fn scc<H: HugrView>(hugr: &H) -> StaticCircuitContext<'_, H> {
    let extensions = ExtensionSet::from_iter([IdentList::new_unchecked("tket2.quantum")]);
    let alloc_ops = hugr
        .nodes()
        .filter(|&x| {
            hugr.get_optype(x)
                .as_extension_op()
                .map_or(false, |ext_op| ext_op.def().name() == "QAlloc")
        })
        .collect();
    let free_ops = hugr
        .nodes()
        .filter(|&x| {
            hugr.get_optype(x)
                .as_extension_op()
                .map_or(false, |ext_op| ext_op.def().name() == "QFree")
        })
        .collect();
    StaticCircuitContext {
        hugr,
        extensions,
        alloc_ops,
        free_ops,
    }
}

fn gates_to_str(gates: Option<Vec<Gate>>) -> String {
    gates.map_or("No static circuit".into(), |gates| {
        gates
            .iter()
            .map(Gate::show)
            .join("\n")
    })
}

#[test]
fn guppy_examples() {
    let pass = StaticCircuitPass::default();
    glob!("static_circuit/guppy_examples/**/*.py", |path| {
        let mut package = T::new(path).run().unwrap();
        let mut reg = STD_REG.to_owned();
        package.update_validate(&mut reg).unwrap();
        let Some([mut hugr]): Option<[Hugr; 1]> = package.modules.try_into().ok() else {
            panic!("Expected exactly one hugr")
        };

        let program_src = {
            let src = fs::read_to_string(path).unwrap();
            let start = src.lines().enumerate().find_map(|(i, line)| line.starts_with("## BEGIN").then_some(i));
            let end = src.lines().enumerate().find_map(|(i, line)| line.starts_with("## END").then_some(i));
            if let (Some(start), Some(end)) = (start,end) {
                assert!(start < end);
                src.lines().skip(start + 1).take(end - start - 1).join("\n")
            } else {
                src
            }
        };

        let mut settings = insta::Settings::clone_current();
        // settings.set_description(program_src);
        settings.set_omit_expression(true);
        settings.bind(|| {
            InlinePass::default().run(&mut hugr, &reg).unwrap();

            let scc = scc(&hugr);
            let scr = StaticCircuitPass::default().run(scc.clone(), &reg).unwrap();
            insta::assert_snapshot!(format!("{program_src}\n===========================\n{}",gates_to_str(scr.static_circuit(scc))));
        });
    });
}
