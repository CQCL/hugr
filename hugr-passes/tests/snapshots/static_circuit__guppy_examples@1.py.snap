---
source: hugr-passes/tests/static_circuit.rs
expression: gates_to_str(gates)
input_file: hugr-passes/tests/static_circuit/guppy_examples/1.py
---
tket2.quantum.Reset [Node(41)]
tket2.quantum.CX [Node(41), Node(53)]
tket2.quantum.Reset [Node(53)]
tket2.quantum.CX [Node(41), Node(53)]
